README.md
# searchpassword
Rapidly check offline if your password is in the compromized passwords list

In order to use this, first obtain the list of hashes of compromized passwords from [https://haveibeenpwned.com/Passwords](https://haveibeenpwned.com/Passwords)

Get the latest one that is the SHA-1 format in order by hash.  Currently, the latest one is Version 8.  It is a 36 GB text file which comes compressed down to a 16 GB 7-zip file for download.  Extract the text file from the downloaded 7-zip file into this project's working directory.

This 36 GB file is pretty large, takes up a lot of disk space, and takes a long time to scan through it looking for a particular password hash, so before you run searchpass.py for the first time, run index_hashfile.py which creates a .yaml file that is an index of the large text file.  It figures out the locations of 4096 segments of the large text file so that later searchpass.py can seek directly to that segment and search through the hash list of only that section.  The index file is in yaml format and takes up 217 kB.  Each of the 4096 segments that it will search through is about 9 MB.

This may be modified to create a larger index and smaller segments to search through, eg. 3.4 MB index for 500 kB segments.  It will take longer to create the index, and will take longer to load the index each time searchpass.py is started, but each password checked will be incredibly fast.  Or reduce the index file to about 13 kB and make each password search through 146 MB worth of hashes.  That will be fairly slow to search for each password.


CHANGELOG.md


## 11/15/22 5:51 AM

- File change history added by Taylor-Jayde Blackstone <t.blackstone@inspyre.tech>
- General readability refactorings;
  - Added constant objects to the module level:
    - INDEX_FILENAME:
        The name of the index file. No path, just name and  extension. String.
    - INDEX_FILEPATH:
        The path of the index file. Includes the file name  and path to parent directory. pathlib.Path object.
        >Note: <br>Also serves as the default index filepath provided  for the **new** :param:`index_file` parameter for the:func:`index_file`.
        
        - HASH_LIST_FILENAME:
            The name of the hash-list file. No path, just name  and extension. String.
        
        - HASH_LIST_FILEPATH:
            The path of the hash-list file. pathlib.Path object.
            
            Also serves as the default hash-list filepath  provided for the **new** :param:``hash_list_file`` 
            parameter.
            
    - Changed names in order to better fit [PEP 8](https://peps.python.org/pep-0008/) conventions:
        |  Old Name   | New Name     |
        |-------------|:-------------|
        |  beginpos   | begin_pos    |
        |  bytecount  | byte_count   |
        | currentdict | current_dict |
        |  hashindex  | hash_index   |
        |   prevpos   | prev_pos     |
        |  searchfor  | query        |
        
        > Note:
            <br>The object previously named `searchfor` wasn't renamed to `search_for`, why for?
        
    - Changed names due to the shadowing of builtin keywords:
        - dict -> _dict
    - Removed names:
        - hashfilehandle
        - indexfilehandle
- Removed file handler references, instead opting for the more  secure, more pythonic 'context manager' pattern.


searchpass.py
#! python3
import yaml   # pip install pyyaml
import hashlib
from getpass import getpass

def indexfile_dict(indexfilename):
    with open(indexfilename,'rt') as indexfilehandle:
        dict = yaml.safe_load(indexfilehandle.read())
    digits = 0
    for item in dict:
        if item[:8].lower() == 'indexes_':
            dict = dict[item][0]
            digits = int(item[8])
    return dict, digits

def ask_password():
    return getpass('Password? (Just press Enter to quit) ')

def sha1sum_password(password):
    hash = hashlib.sha1(password.encode('utf-8'))
    return hash.hexdigest().upper()

def segment(dict, sha, digits):
    fewdigits = sha[:digits]
    begin = dict[fewdigits]["begin_byte"]
    length = dict[fewdigits]["byte_count"]

    return begin, length

def read_segment(hashfilehandle, sha, begin, length, digits):
    hashfilehandle.seek(begin, 0)
    end = begin + length
    pos = hashfilehandle.tell()
    match = False
    while pos < end and not match:
        line = hashfilehandle.readline()
        pos = hashfilehandle.tell()
        if sha[:digits].upper() != line[:digits].upper():
            print(
                f"Error: Need to run index_hashfile.py or there is a bug. Searching for {sha[:digits].upper()} on line: "
            )

            print(line)
            break
        if sha.upper() == line[:40].upper():
            match = True
            print("Match found:")
            print(line)
            break
    if not match:
        print("Match not found.\n")

def main():
    indexfilename = 'pwned-passwords-hash-index_v8.yaml'
    hashlistfilename = 'pwned-passwords-sha1-ordered-by-hash-v8.txt'
    with open(hashlistfilename,'rt') as hashfilehandle:
        print(f"Reading {indexfilename}")
        dict = indexfile_dict(indexfilename)
        digits = dict[1]
        dict = dict[0]
        # print(dict)
        print("")

        password = " "
        while password != "":
            password = ask_password()
            if len(password) < 1:
                break
            sha = sha1sum_password(password)
                    # password = ""
            print(f"Looking for SHA1SUM: {sha}")
            begin = segment(dict, sha, digits)[0]
            length = segment(dict, sha, digits)[1]
            read_segment(hashfilehandle, sha, begin, length, digits)

if __name__ == "__main__":
    main()


index_hashfile.py
#!/usr/bin/env python3

from __future__ import annotations
from pathlib import Path
import yaml  # pip install pyyaml
from dill import dump
from tqdm import trange


FILE_DIR = Path(__file__).parent

INDEX_FILENAME = 'pwned-passwords-hash-index_v8.yaml'
INDEX_FILEPATH = FILE_DIR.joinpath(INDEX_FILENAME)

HASH_LIST_FILENAME = 'pwned-passwords-sha1-ordered-by-hash-v8.txt'
HASH_LIST_FILEPATH = FILE_DIR.joinpath(HASH_LIST_FILENAME)


def build_string(digits):
    return f'%0{str(digits)}X'


def index_file(
        digits,
        index_fp: (str | Path) = INDEX_FILEPATH,
        hash_list_file: (str | Path) = HASH_LIST_FILEPATH
):
    index_fp = index_fp
    hash_list_fp = hash_list_file

    with open(hash_list_fp, 'rt') as hf:
        begin = []
        byte_count = []

        for hash_index in trange(16 ** digits):

            query = build_string(digits) % hash_index
            # print(query)
            begin_pos = hf.tell()
            prev_pos = begin_pos
            current = hf.readline()[:digits]
            while query == current:
                # print(searchfor + " " + current)
                # print("  " + str(hashfilehandle.tell()))
                prev_pos = hf.tell()
                current = hf.readline()[:digits]

            begin.append(begin_pos)
            byte_count.append(prev_pos - begin_pos)

        # print("\n\n")
        _dict = {
                f'indexes_{str(digits)}_digit_hex': []
        }
        current_dict = {
                build_string(digits) % index_num: {
                        "begin_byte": begin[index_num],
                        "byte_count": byte_count[index_num]
                } for index_num in range(len(begin))
        }

        _dict[f'indexes_{str(digits)}_digit_hex'].append(current_dict)
        # indexfilehandle.write(yaml.dump(dict,sort_keys=False))

    with open(index_fp, 'w') as ifp:
        ifp.write(yaml.dump(_dict))

    with open(Path('./index.pkl').resolve(), 'wb') as file:
        dump(_dict, file)


def main():
    digits = 3
    print(f"File will be split {str(16 ** digits)} ways.")
    index_file(digits)


if __name__ == "__main__":
    main()

"""
File Change History:

    - 11/15/22 5:51 AM:
        - File change history added by Taylor-Jayde Blackstone <t.blackstone@inspyre.tech>
        - General readability refactorings;
            - Added constant objects to the module level:
                - INDEX_FILENAME:
                    The name of the index file. No path, just name and extension. String.

                - INDEX_FILEPATH:
                    The path of the index file. Includes the file name and path to parent directory. pathlib.Path 
                    object.

                    Also serves as the default index filepath provided for the **new** :param:`index_file` parameter.

                - HASH_LIST_FILENAME:
                    The name of the hash-list file. No path, just name and extension. String.

                - HASH_LIST_FILEPATH:
                    The path of the hash-list file. pathlib.Path object.

                    Also serves as the default hash-list filepath provided for the **new** :param:'hash_list_file` 
                    parameter.

            - Changed names in order to better fit (PEP 8)[https://peps.python.org/pep-0008/] conventions:
                - bytecount -> byte_count
                - hashindex -> hash_index
                - searchfor -> query 
                - beginpos -> begin_pos
                - prevpos -> prev_pos
                - currentdict -> current_dict
            - Changed names due to the shadowing of builtin keywords:
                - dict -> _dict
            - Removed names:
                - hashfilehandle -> Replaced with context manager
                - indexfilehandle -> Replaced with context manager
        - Removed file handler references, instead opting for the more secure, more pythonic 'context manager' pattern.

"""
